packages/gatsby-source-drupal/src/gatsby-node.js

```diff
@@ -102,7 +104,96 @@ exports.sourceNodes = async (
     return
   }
 
-  const drupalFetchActivity = reporter.activityTimer(`Fetch data from Drupal`)
+  fastBuilds = fastBuilds || false
+  if (fastBuilds) {
+    let lastFetched = 0
+    if (
+      store.getState().status.plugins &&
+      store.getState().status.plugins[`gatsby-source-drupal`] &&
+      store.getState().status.plugins[`gatsby-source-drupal`].lastFetched
+    ) {
+      lastFetched = store.getState().status.plugins[`gatsby-source-drupal`]
+        .lastFetched
+    }
+
+    const drupalFetchIncrementalActivity = reporter.activityTimer(
+      `Fetch incremental changes from Drupal`
+    )
+    let requireFullRebuild = false
+
+    drupalFetchIncrementalActivity.start()
+
+    try {
+      // Hit fastbuilds endpoint with the lastFetched date.
+      const data = await axios.get(
+        `${baseUrl}/gatsby-fastbuilds/sync/${lastFetched}`,
+        {
+          auth: basicAuth,
+          headers,
+          params,
+        }
+      )
+
+      if (data.data.status === -1) {
+        reporter.info(`Unable to pull incremental data changes from Drupal`)
+        setPluginStatus({ lastFetched: data.data.timestamp })
+        requireFullRebuild = true
+      } else {
+        // Touch nodes so they are not garbage collected by Gatsby.
+        const existingNodes = getNodes().filter(
+          n => n.internal.owner === `gatsby-source-drupal`
+        )
+        existingNodes.forEach(n => touchNode({ nodeId: n.id }))
+
+        // Process sync data from Drupal.
+        let nodesToSync = data.data.entities
+        for (const nodeSyncData of nodesToSync) {
```

Comment:
I'm a little weiry of loop-in-a-loop. Even if this is incremental, could this worst case still pull in a million nodes? What happens at a cold build? This might just be the way it is but can you double check to confirm that the performance scales linearly with the number of nodes?

Reply:
The way the the data is sent back from Drupal does require this. For instance, if a piece of content is updated but it also contains an image and possibly other related entities, it would need to update all of these nodes. So while there is a nested loop here, each node needs to be updated.

There is logic on the Drupal site to prevent nodes from being updated multiple times. For instance, if a piece of content is updated three times, only the most recent change is sent to Gatsby.

At a cold build this logic is not run and instead a full build is initiated and the [code here](https://github.com/gatsbyjs/gatsby/blob/caea9adff18c169b1521109da653acf25c87dfd1/packages/gatsby-source-drupal/src/gatsby-node.js#L211) is executed.